### ISSUE TITLE
False negative when `match`ing a Protocol

### ISSUE THREAD

**Issue by lihu-zhong** [astral-sh/ty#2581]
### Summary

**Observed behavior:**

`ty` correctly raises an `invalid-argument-type` error when using a `Protocol` in an `isinstance()` check, but doesn't raise the same error when an equivalent runtime check is done with `match`:

```python
from dataclasses import dataclass
from typing import Any, Protocol


class Foo(Protocol):
    foo: str


@dataclass
class ImplFoo:
    foo: str = "bar"


def isinstance_check(input: Any) -> None:
    if isinstance(input, Foo):
        input.foo


def match_check(input: Any) -> None:
    match input:
        case Foo():
            input.foo


isinstance_check(ImplFoo())
match_check(ImplFoo())
```
In the above code, `ty` raises one error.

**Expected Behavior:**

`ty` should raise the same error on the `match` structure, since it raises the same runtime error. `mypy` misses this as well, but `pyright` catches it.

### Version

0.0.10

**Comment by AlexWaygood** [comment]
Thanks! Yes, it seems reasonable to expand the rule so that it catches this runtime error. Mypy and pyrefly have the same false negative that we do here, but pyright appears to catch this.

Edit: oops, i see you already analysed what other type checkers do here — thanks!

**Comment by denyszhak** [comment]
@AlexWaygood May I take on this one?

**Comment by carljm** [comment]
@denyszhak Certainly!
